File: solve.cpp
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */

#include<iostream>
#include<string>
#include<vector>
#include<algorithm>
#include<map>
#include<stack>
#include<queue>

using namespace std;

//显示中间结果
void print(vector<int> a)
{
    for(int i=0;i<(int)a.size();i++)
        cout<<a[i]<<" ";
    cout<<endl;
}

void print(vector<vector<int>> a)
{
    for(int i=0;i<(int)a.size();i++)
    {
        for(int j=0;j<(int)a[i].size();j++)
            cout<<a[i][j]<<" ";
        cout<<endl;
    }
}

void print(vector<vector<char>> a)
{
    for(int i=0;i<(int)a.size();i++)
    {
        for(int j=0;j<(int)a[i].size();j++)
            cout<<a[i][j]<<" ";
        cout<<endl;
    }
}

void print(vector<bool> a)
{
    for(int i=0;i<(int)a.size();i++)
    {
        cout<<i<<" ";
        if(a[i])
            cout<<"true ";
        else
            cout<<"false ";
    }
    cout<<endl;
}

void print(bool a[],int len)
{
    for(int i=0;i<len;i++)
    {
        cout<<i<<" ";
        if(a[i])
            cout<<"true ";
        else
            cout<<"false ";
    }
    cout<<endl;
}

//#define testMod

#ifdef testMod
void test()
{
    
}
#endif

#ifndef testMod
class Solution {
private:
    vector<vector<char>> grid;
public:
    void solve(vector<vector<char>>& board) {
        int count=0;
        this->grid=board;
        for(int i=0;i<(int)grid.size();i++)
        {
            for(int j=0;j<(int)grid[0].size();j++)
            {
                if(this->grid[i][j]=='O')
                {
                    if(!find(i,j))
                        fill(i,j);
                    count++;
                }
            }
        }
        for(int i=0;i<(int)grid.size();i++)
        {
            for(int j=0;j<(int)grid[0].size();j++)
            {
                if(this->grid[i][j]=='X')
                {
                    board[i][j]='X';
                }
                else
                    board[i][j]='O';
            }
        }
    }

    int find(int i,int j)
    {
        if(i<0||j<0)
            return 1;
        if(i>=grid.size()||j>=grid[0].size())
            return 1;
        if(grid[i][j]!='O')
            return 0;
        cout<<"i "<<i<<" j "<<j<<endl;
        grid[i][j]='1';
        print(grid);
        cout<<endl;
        return find(i-1,j)+
        find(i+1,j)+
        find(i,j+1)+
        find(i,j-1);
    }

    void fill(int i,int j)
    {
        if(i<0||j<0)
            return;
        if(i>=grid.size()||j>=grid[0].size())
            return;
        if(grid[i][j]=='X')
            return;
        grid[i][j]='X';
        fill(i-1,j);
        fill(i+1,j);
        fill(i,j+1);
        fill(i,j-1);
    }
};
#endif

int main()
{
#ifdef testMod
    test();
#endif

#ifndef testMod
    Solution sl;
    vector<vector<char>> in=
    {{'X','X','X','X'},
        {'X','O','O','X'},
        {'X','X','O','X'},
        {'X','O','X','X'}};
    print(in);
    sl.solve(in);
    print(in);

#endif 
    return 0;
}

File: solve.py
from typing import *

class Solution:
    def solve(self, board: List[List[str]]) -> None:
        self.board=board
        self.x=len(board)
        if self.x==0:
            return 
        self.y=len(board[0])
        if self.y==0:
            return
        for i in range(self.x):
            if board[i][0]=='O':
                self.fill(i,0)
            if board[i][-1]=='O':
                self.fill(i,self.y-1)
        for j in range(self.y):
            if board[0][j]=='O':
                self.fill(0,j)
            if board[-1][j]=='O':
                self.fill(self.x-1,j)
        def convert(a):
            if a=='X' or a=='O':
                return 'X'
            return 'O'
        for i in range(self.x):
            for j in range(self.y):
                board[i][j]=convert(board[i][j])
                
    def fill(self,i,j):
        if i<0 or j<0 or i>=self.x or j>=self.y:
            return 
        if self.board[i][j]!='O':
            return 
        self.board[i][j]='1'
        self.fill(i-1,j)
        self.fill(i+1,j)
        self.fill(i,j-1)
        self.fill(i,j+1)
sl=Solution()
board=[['X','X','X','X'], ['X','O','O','X'], ['X','X','O','X'], ['X','O','X','X']] 
sl.solve(board)
print(board)

